poetry/puzzle/provider.py

```diff
@@ -67,6 +68,11 @@ def __init__(
         self._in_progress = False
         self._overrides = {}
         self._deferred_cache = {}
+        self._locked_packages = {}
```

Comment:
I am not sure this is the right place to put this.

We already have a locked system in the `VersionSolver` class, so we might be able to fix and improve it there.

Reply:
I'll have a whack at that; I was going for the lowest possible place to put this logic otherwise we will end up re-implementing the logic at multiple places. It might also be a good idea to re-work the structure a bit for `2.x`.